File: Matrix/Java/Battleships in a Board.java
// Problem link: https://leetcode.com/problems/battleships-in-a-board/description/
// Time: O(n * m)
// Space: O(1)
class Solution {
    public int countBattleships(char[][] board) {
        int rows = board.length, cols = board[0].length, result = 0;

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(board[row][col] == 'X') {
                    this.dfs(row, col, board, rows, cols);
                    result += 1;
                }
            }
        }

        return result;
    }

    private void dfs(int row, int col, char[][] board, int rows, int cols) {
        if(row < 0 || col < 0 || row == rows || col == cols) {
            return;
        }

        board[row][col] = '.';

        if(row + 1 < rows && board[row + 1][col] == 'X') {
            this.dfs(row + 1, col, board, rows, cols);
        }
        else if(col + 1 < cols && board[row][col + 1] == 'X') {
            this.dfs(row, col + 1, board, rows, cols);
        }
    }
}

File: Matrix/Python/Battleships in a Board.py
# Problem link: 
# Time: O(n * m)
# Space: O(1)
class Solution:
    def countBattleships(self, board: List[List[str]]) -> int:
        rows, cols, result = len(board), len(board[0]), 0
       

        for row in range(rows):
            for col in range(cols):
                if board[row][col] == 'X':
                    self.dfs(row, col, board, rows, cols)
                    result += 1
        
        return result

    def dfs(self, row, col, board, rows, cols):
        if row < 0 or col < 0 or row == rows or col == cols:
            return
        
        board[row][col] = '.'

        if row + 1 < rows and board[row + 1][col] == 'X':
            self.dfs(row + 1, col, board, rows, cols)
        elif col + 1 < cols and board[row][col + 1] == 'X':
            self.dfs(row, col + 1, board, rows, cols)
